clean_dockerhub.py

```diff
@@ -0,0 +1,277 @@
+                # Rate limit hit
+                raise TooManyRequestsError(f"{namespace}/{repo}:{tag}")
+            return resp.status
+
+
+async def get_old_tags(
+    namespace: str,
+    repo: str,
+    token: str,
+    min_age_in_days: int,
+    max_number_of_pages: int,
+) -> list[Tag]:
+    """
+    Calculates the time threshold, opens a aiohttp.ClientSession session and then calls `fetch_old_tags`.
+    """
+    threshold = datetime.now() - timedelta(days=min_age_in_days)
+
+    conn = aiohttp.TCPConnector(limit_per_host=10)
+    async with aiohttp.ClientSession(connector=conn) as session:
+        # 2) Fetch all tags
+        old_tags = await fetch_old_tags(
+            session, namespace, repo, threshold, token, max_number_of_pages
+        )
+        return old_tags
+
+
+async def delete_tags(
+    namespace: str,
+    repo: str,
+    token: str,
+    tags_to_delete: list[Tag],
+) -> bool:
+    """
+    Delete given tags from a repository, always with 3 Http requests in parallel.
+    The method creates an aiohttp.ClientSession session, then creates all async delete tasks.
+    Whenever one tasks raises the TooManyRequestsError exception, all other tasks are cancelled,
+    and the method returns True. If no such exception is raised, the method returns False.
+    """
+
+    # Higher values will run deletion faster, but have higher risk of getting a 429,
+    # see https://docs.docker.com/docker-hub/usage/#abuse-rate-limit
+    # Tests have shown that max deletion rate without running into rate limits is ~650 Tags/min.
+    sem = asyncio.Semaphore(2)
+
+    retry = False
+    conn = aiohttp.TCPConnector(limit_per_host=10)
+    async with aiohttp.ClientSession(connector=conn) as session:
+        try:
+            tasks = [
+                asyncio.create_task(
+                    delete_tag(session, namespace, repo, tag, token, sem)
+                )
+                for tag in tags_to_delete
+            ]
+
+            status_counter_map: dict[int, int] = {}
+            retry = False
+
+            for delete_coro in tqdm(
+                asyncio.as_completed(tasks),
+                total=len(tasks),
+                desc="Deleting old tags",
+            ):
+                try:
+                    status = await delete_coro
+                    status_counter_map[status] = status_counter_map.get(status, 0) + 1
+                except TooManyRequestsError as e:
+                    # upon 429, cancel all pending tasks
+                    retry = True
+                    for t in tasks:
+                        if not t.done():
+                            t.cancel()
+                    break
+
+        finally:
+            # ensure all tasks are concluded to suppress warnings
+            await asyncio.gather(*tasks, return_exceptions=True)
+
+        print(
+            f"Deletion run completed. \n Status counters: \n {status_counter_map}\n Retry: {retry}",
+        )
+    return retry
+
+
+async def clean_dockerhub(
```

Comment:
As the module is called `clean_dockerhub`, this function maybe could be called `clean_repository` or `clean_docker_repository`. 

Also, rather than a generic name like "clean" maybe something like "delete old tags" would be more precise and avoid creating wrong expectations.

Reply:
Changed the name to "fetch_and_delete_old_tags"